src/precice/impl/WriteDataContext.cpp

```diff
@@ -9,29 +11,104 @@ WriteDataContext::WriteDataContext(
     mesh::PtrMesh mesh)
     : DataContext(data, mesh)
 {
+  _writeDataBuffer = time::Sample{Eigen::VectorXd(), Eigen::MatrixXd()};
 }
 
-void WriteDataContext::writeValues(::precice::span<const VertexID> vertices, ::precice::span<const double> values)
+void WriteDataContext::resetData(bool atEndOfWindow)
 {
+  // See also https://github.com/precice/precice/issues/1156.
+  _providedData->toZero();
+
+  // reset writeDataBuffer
+  _writeDataBuffer.values.setZero();
+  _writeDataBuffer.gradients.setZero();
+
+  PRECICE_ASSERT(!hasReadMapping(), "Read mapping is not allowed for WriteDataContext.");
+  // reset all toData
+  if (hasWriteMapping()) {
+    std::for_each(_mappingContexts.begin(), _mappingContexts.end(), [](auto &context) { context.toData->toZero(); });
+  }
+
+  if (atEndOfWindow) {
+    // need to not only clear _providedData->timeStepsStorage(), but also _toData->timeStepsStorage() as soon as we map from storage to storage.
+    _providedData->timeStepsStorage().clear();
+  }
+}
+
+void WriteDataContext::writeValuesIntoDataBuffer(::precice::span<const VertexID> vertices, ::precice::span<const double> values)
+{
+  PRECICE_ASSERT(vertices.size() * getDataDimensions() == values.size());
```

Comment:
This assertion is also checked in ParticipantImp::writeData, looks like it can be removed from there